aten/src/ATen/native/GridSampler.cpp

```diff
@@ -0,0 +1,780 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/detail/CUDAHooksInterface.h"
+#include "ATen/native/GridSampler.h"
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace at { namespace native {
+
+using at::native::detail::GridSamplerInterpolation;
+using at::native::detail::GridSamplerPadding;
+
+namespace {
+  static inline int64_t clip_coordinates(int64_t in, int64_t clip_limit) {
+    return std::min(clip_limit - 1, std::max(in, static_cast<int64_t>(0)));
+  }
+
+  static inline bool within_bounds_2d(int64_t h, int64_t w, int64_t H, int64_t W) {
+    return h >= 0 && h < H && w >= 0 && w < W;
+  }
+
+  static inline bool within_bounds_3d(int64_t d, int64_t h, int64_t w, int64_t D, int64_t H, int64_t W) {
+    return d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W;
+  }
+
+  template<typename scalar_t>
+  static inline void safe_add_2d(scalar_t *data, int64_t h, int64_t w,
+                                 int64_t sH, int64_t sW, int64_t H, int64_t W,
+                                 scalar_t delta) {
+    if (h >= 0 && h < H && w >= 0 && w < W) {
+      data[h * sH + w * sW] += delta;
+    }
+  }
+
+  template<typename scalar_t>
+  static inline void safe_add_3d(scalar_t *data, int64_t d, int64_t h, int64_t w,
+                                 int64_t sD, int64_t sH, int64_t sW,
+                                 int64_t D, int64_t H, int64_t W,
+                                 scalar_t delta) {
+    if (d >=0 && d < D && h >= 0 && h < H && w >= 0 && w < W) {
+      data[d * sD + h * sH + w * sW] += delta;
+    }
+  }
+
+  template<typename scalar_t>
+  Tensor grid_sampler2d_cpu_impl(const Tensor& input, const Tensor& grid,
+                                 GridSamplerInterpolation interpolation_mode,
+                                 GridSamplerPadding padding_mode) {
+    int64_t N = input.size(0);
+    int64_t C = input.size(1);
+    int64_t inp_H = input.size(2);
```

Comment:
These assignments... are really annoying haha. This is probably because in the C10 refactoring we regressed fastGet/fastSet.

Reply:
Is there any reason you couldn't have used an ATen accessor here?

Reply:
No particular reason, but I liked working with pointers here because the loops are easy to work with to optimize a lot of redundant computation on offset (pointer) to get/set values.